src/config.rs

```diff
@@ -1,14 +1,87 @@
-pub use iotics_identity::{create_agent_auth_token, Config, IdentityLibError};
+use std::sync::{Arc, Mutex};
+
+use iotics_grpc_client::auth_builder::IntoAuthBuilder;
+use iotics_identity::{create_agent_auth_token, Config};
 
 use crate::constants::AGENT_KEY_NAME;
 
 #[derive(Debug, Clone)]
-pub struct ApiConfig {
+pub struct AuthBuilder {
+    api_config: Arc<Mutex<ApiConfig>>,
+    token: Arc<Mutex<Option<String>>>,
+}
+
+impl AuthBuilder {
+    pub fn new() -> Arc<Self> {
+        let api_config = get_api_config();
+
+        Arc::new(Self {
+            api_config: Arc::new(Mutex::new(api_config)),
+            token: Arc::new(Mutex::new(None)),
+        })
+    }
+
+    pub fn get_identity_config(&self) -> Result<Config, anyhow::Error> {
+        let api_config_lock = self
+            .api_config
+            .lock()
+            .map_err(|_| anyhow::anyhow!("failed to lock the settings mutex"))?;
+
+        Ok(api_config_lock.identity_config.clone())
+    }
+}
+
+impl IntoAuthBuilder for AuthBuilder {
+    fn get_host(&self) -> Result<String, anyhow::Error> {
+        let api_config_lock = self
+            .api_config
+            .lock()
+            .map_err(|_| anyhow::anyhow!("failed to lock the settings mutex"))?;
+
+        Ok(api_config_lock.host_address.clone())
+    }
+
+    fn get_token(&self) -> Result<String, anyhow::Error> {
+        let mut token_lock = self
+            .token
+            .lock()
+            .map_err(|_| anyhow::anyhow!("failed to lock the token mutex"))?;
+
+        if token_lock.is_none() {
+            let api_config_lock = self
+                .api_config
+                .lock()
+                .map_err(|_| anyhow::anyhow!("failed to lock the settings mutex"))?;
+
+            let identity_config = Config {
+                resolver_address: api_config_lock.identity_config.resolver_address.clone(),
+                token_duration: api_config_lock.identity_config.token_duration as i64,
+                user_did: api_config_lock.identity_config.user_did.clone(),
+                agent_did: api_config_lock.identity_config.agent_did.clone(),
+                agent_key_name: api_config_lock.identity_config.agent_key_name.clone(),
+                agent_name: api_config_lock.identity_config.agent_name.clone(),
+                agent_secret: api_config_lock.identity_config.agent_secret.clone(),
+            };
+
+            let token = create_agent_auth_token(&identity_config)?;
+            let token = format!("bearer {}", token);
```

Comment:
Is your syntax always preferable over the following ? I find this a little weird
```rust
let mut token = create_agent_auth_token(&identity_config)?;
token = format!("bearer {}", token);
```
[here](http://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/variable-bindings.html#mutability) it is mentioned about it

Reply:
The shadowing achieved by using the same variable name is mostly a matter of preference, I guess. I like the fact that it makes it obvious that the previous `token` goes out of scope after the re-declaration.

I prefer the current implementing over the suggestion because it keeps the binding immutable, which is really what we want.

A mutable binding might mislead the developer down the line that token is something that they are able to mutate, when in reality it isn't.
Knowing that something is immutable also helps the compiler to apply all the possible optimizations.